iml-services/iml-device/src/main.rs

```diff
@@ -45,6 +50,140 @@ async fn create_cache(pool: &DbPool) -> Result<Cache, ImlDeviceError> {
+        .await;
+    match new_device {
+        Ok(d) => {
+            tracing::info!("Inserted devices from host '{}'", host);
+            tracing::debug!("Inserted {:?}", d);
+        }
+        Err(e) => {
+            tracing::error!("Error saving new devices: {}", e);
+        }
+    }
+}
+
+async fn update_client_mounts(
+    pool: &DbPool,
+    ct_id: Option<i32>,
+    host: &Fqdn,
+    mounts: &HashSet<Mount>,
+) {
+    let host_id = match host::table
+        .select(host::id)
+        .filter(host::fqdn.eq(host.to_string()))
+        .get_result_async::<i32>(&pool)
+        .await
+    {
+        Ok(id) => {
+            tracing::debug!("Host id: {}", id);
+            id
+        }
+        Err(e) => {
+            tracing::error!("Host '{}' is unknown: {}", host, e);
+            return;
+        }
+    };
+
+    let lustre_mounts = mounts
+        .into_iter()
+        .filter(|m| m.fs_type.0.as_str() == "lustre")
+        .filter_map(|m| {
+            m.source
+                .0
+                .to_str()
+                .and_then(|p| p.splitn(2, ":/").nth(1))
+                .map(|fs| (fs.to_string(), m.target.0.to_str().map(String::from)))
+        })
+        .collect::<Vec<_>>();
+
+    tracing::debug!("Client mounts at {}: {:?}", host, &lustre_mounts);
+
+    let state_modified_at = chrono::offset::Utc::now().into_sql::<sql_types::Timestamptz>();
+
+    let insert_mounts = lustre_mounts
+        .iter()
+        .map(|(fs, tg)| {
+            let mp = tg
+                .clone()
+                .into_sql::<sql_types::Nullable<sql_types::Text>>();
+            fs::table
+                .select((
+                    host_id.into_sql::<sql_types::Integer>(),
+                    fs::id,
+                    mp,
+                    "mounted".as_sql::<sql_types::Text>(),
+                    state_modified_at,
+                    false.into_sql::<sql_types::Bool>(),
+                    Some(true).into_sql(),
+                    ct_id.into_sql(),
+                ))
+                .filter(fs::name.eq(fs.clone()))
+                .insert_into(clmnt::table)
+                .into_columns((
+                    clmnt::host_id,
+                    clmnt::filesystem_id,
+                    clmnt::mountpoint,
+                    clmnt::state,
+                    clmnt::state_modified_at,
+                    clmnt::immutable_state,
+                    clmnt::not_deleted,
+                    clmnt::content_type_id,
+                ))
+                .on_conflict((clmnt::host_id, clmnt::filesystem_id, clmnt::not_deleted))
+                .do_update()
+                .set((
+                    clmnt::mountpoint.eq(excluded(clmnt::mountpoint)),
+                    clmnt::state.eq(excluded(clmnt::state)),
+                    clmnt::state_modified_at.eq(state_modified_at),
+                ))
+                .get_result_async::<ChromaCoreLustreclientmount>(&pool)
+        })
+        .collect::<Vec<_>>();
```

Comment:
This still does one query + 1 subselect for each incoming mount. The code I posted earlier does a single batch upsert, which is preferable to this approach.

Reply:
Yeah, but we still need to set the foreign key.

Reply:
I think we're going to have to get rid of the foreign key constraint first.

This patch won't be reliable without that change, because any mounts that come in before fs discovery will be discarded.